### PR TITLE
add GMS 1.4+ function buffer_get_address()

### PR DESCRIPTION
**Pull request by time-killer-games**
Josh doesn't approve of yet another good feature yoyogames provides, no excuse why we don't have this other than Josh saying baselessly "it's not a good idea".... A lot of people will need this to take elements provided from dll's and have them rendered to a surface, like my video player, or a chomium-based web browser, and many other good things we can't add without it. Unless we want to break GM compatibity, of course. I thought he didn't like doing that. Make up your mind, Josh!

At any rate, not even sure if i did this right.

### CODE REVIEW

**Comment by time-killer-games** [comment]
@JoshDreamland now passing joo bitch!!!

**Comment by time-killer-games** [comment]
@JoshDreamland as you can see, this project now builds:
[WebMPlayer.gmx.zip](https://github.com/enigma-dev/enigma-dev/files/6370569/WebMPlayer.gmx.zip)
Although the visual part of the video needs to be done still. If you don't have Ubuntu 20.04 binary compatibility, you will need to build the *.so from source code provided.

File: ENIGMAsystem/SHELL/Universal_System/bufferstruct.cpp [hunk 1]
```diff
@@ -234,6 +234,17 @@ void buffer_fill(int buffer, unsigned offset, int type, variant value, unsigned
...
+  
+void *buffer_get_address(int buffer) {
+  #ifdef DEBUG_MODE
+  if (buffer < 0 or size_t(buffer) >= enigma::buffers.size() or !enigma::buffers[buffer]) {
+    DEBUG_MESSAGE("Attempting to access non-existing buffer " + toString(buffer), MESSAGE_TYPE::M_USER_ERROR);
+    return nullptr;
+  }
+  #endif
+  enigma::BinaryBuffer *binbuff = enigma::buffers[buffer];
+  return reinterpret_cast<void *>(&binbuff->data[0]);
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/bufferstruct.cpp:246]: If this is a vector, please favor `vec.data()` over `&vec[0]`.

**Review by JoshDreamland — APPROVED** [review]

File: ENIGMAsystem/SHELL/Universal_System/bufferstruct.cpp (hunk 1 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/bufferstruct.cpp]: aye aye, captain!

File: ENIGMAsystem/SHELL/Universal_System/bufferstruct.cpp [hunk 2]
```diff
@@ -256,6 +267,12 @@ void buffer_get_surface(int buffer, int surface, int mode, unsigned offset, int
   //TODO: Write this function
 }
 
+//NOTE: This function should most likely be added in graphics systems.
+void buffer_set_surface(int buffer, int surface, int mode, unsigned offset, int modulo) {
+  //get_buffer(binbuff, buffer);
```
Comment (fundies) [inline ENIGMAsystem/SHELL/Universal_System/bufferstruct.cpp:272]: If you're going to put a stub it's a good habit to put DEBUG_MESSAGE("Unimplemented", M_WARNING) or whatever in it to lessen confusion when people's games call it and it doesn't work.

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/bufferstruct.cpp:272]: on it as we speak, thanks!

**Pull request merged**